File: skills/agent-patterns/templates/c/routing.c
/**
 * Routing Pattern Implementation for C
 * Classification-based routing of inputs to specialized handlers
 *
 * Note: This is a simplified example. In production, use a proper HTTP library
 * like libcurl and JSON library like cJSON or jansson.
 *
 * Compile with:
 * gcc -o routing routing.c -lcurl -ljson-c
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// Maximum sizes
#define MAX_CATEGORIES 20
#define MAX_INPUT_SIZE 4096
#define MAX_OUTPUT_SIZE 16384
#define MAX_CATEGORY_NAME 64
#define MAX_DESCRIPTION 256

// Forward declarations
typedef struct ClassificationResult ClassificationResult;
typedef struct Route Route;
typedef struct Router Router;

/**
 * Classification result from LLM
 */
typedef struct ClassificationResult {
    char category[MAX_CATEGORY_NAME];
    double confidence;
    char reasoning[MAX_DESCRIPTION];
} ClassificationResult;

/**
 * Route handler function type
 */
typedef char* (*RouteHandler)(const char* input, void* user_data);

/**
 * Route definition
 */
typedef struct Route {
    char category[MAX_CATEGORY_NAME];
    char description[MAX_DESCRIPTION];
    RouteHandler handler;
    void* user_data;
} Route;

/**
 * Router that classifies and routes inputs
 */
typedef struct Router {
    char* api_key;
    char* model;
    Route routes[MAX_CATEGORIES];
    size_t route_count;
    double confidence_threshold;
    RouteHandler fallback_handler;
    void* fallback_user_data;
} Router;

/**
 * Simplified API call - in production, use libcurl
 */
char* call_anthropic_api(const char* api_key, const char* model,
                         const char* prompt, int max_tokens) {
    // NOTE: This is a placeholder. In production, implement using libcurl:
    //
    // Example with libcurl:
    // CURL *curl = curl_easy_init();
    // struct curl_slist *headers = NULL;
    // headers = curl_slist_append(headers, "Content-Type: application/json");
    // headers = curl_slist_append(headers, api_key_header);
    // headers = curl_slist_append(headers, "anthropic-version: 2023-06-01");
    // curl_easy_setopt(curl, CURLOPT_URL, "https://api.anthropic.com/v1/messages");
    // curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    // curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_body);
    // ...

    printf("API Call (mock):\n");
    printf("Model: %s\n", model);
    printf("Max tokens: %d\n", max_tokens);
    printf("Prompt: %.100s...\n", prompt);

    // Mock response - return classification JSON
    return strdup("{\"category\": \"general\", \"confidence\": 0.85, \"reasoning\": \"Mock classification\"}");
}

/**
 * Parse classification JSON response
 */
bool parse_classification(const char* json, ClassificationResult* result) {
    // Simplified JSON parsing - use cJSON or jansson in production

    // Find category
    const char* cat_start = strstr(json, "\"category\"");
    if (cat_start) {
        cat_start = strchr(cat_start + 10, '"');
        if (cat_start) {
            cat_start++;
            const char* cat_end = strchr(cat_start, '"');
            if (cat_end) {
                size_t len = cat_end - cat_start;
                if (len >= MAX_CATEGORY_NAME) len = MAX_CATEGORY_NAME - 1;
                strncpy(result->category, cat_start, len);
                result->category[len] = '\0';
            }
        }
    }

    // Find confidence
    const char* conf_start = strstr(json, "\"confidence\"");
    if (conf_start) {
        conf_start = strchr(conf_start + 12, ':');
        if (conf_start) {
            result->confidence = atof(conf_start + 1);
        }
    }

    // Find reasoning
    const char* reason_start = strstr(json, "\"reasoning\"");
    if (reason_start) {
        reason_start = strchr(reason_start + 11, '"');
        if (reason_start) {
            reason_start++;
            const char* reason_end = strchr(reason_start, '"');
            if (reason_end) {
                size_t len = reason_end - reason_start;
                if (len >= MAX_DESCRIPTION) len = MAX_DESCRIPTION - 1;
                strncpy(result->reasoning, reason_start, len);
                result->reasoning[len] = '\0';
            }
        }
    }

    return strlen(result->category) > 0;
}

/**
 * Create a new router
 */
Router* router_create(const char* api_key, const char* model) {
    Router* router = (Router*)calloc(1, sizeof(Router));
    router->api_key = strdup(api_key);
    router->model = model ? strdup(model) : strdup("claude-sonnet-4-20250514");
    router->route_count = 0;
    router->confidence_threshold = 0.7;
    router->fallback_handler = NULL;
    return router;
}

/**
 * Set confidence threshold
 */
void router_set_threshold(Router* router, double threshold) {
    router->confidence_threshold = threshold;
}

/**
 * Set fallback handler
 */
void router_set_fallback(Router* router, RouteHandler handler, void* user_data) {
    router->fallback_handler = handler;
    router->fallback_user_data = user_data;
}

/**
 * Add a route to the router
 */
bool router_add_route(Router* router, const char* category,
                      const char* description, RouteHandler handler,
                      void* user_data) {
    if (router->route_count >= MAX_CATEGORIES) {
        return false;
    }

    Route* route = &router->routes[router->route_count];
    strncpy(route->category, category, MAX_CATEGORY_NAME - 1);
    strncpy(route->description, description, MAX_DESCRIPTION - 1);
    route->handler = handler;
    route->user_data = user_data;
    router->route_count++;

    return true;
}

/**
 * Classify an input
 */
ClassificationResult* router_classify(Router* router, const char* input) {
    // Build classification prompt
    char prompt[MAX_INPUT_SIZE * 2];
    char categories[MAX_INPUT_SIZE];
    categories[0] = '\0';

    for (size_t i = 0; i < router->route_count; i++) {
        char cat_line[MAX_CATEGORY_NAME + MAX_DESCRIPTION + 10];
        snprintf(cat_line, sizeof(cat_line), "%s: %s\n",
                 router->routes[i].category, router->routes[i].description);
        strcat(categories, cat_line);
    }

    snprintf(prompt, sizeof(prompt),
        "Classify the following input into one of these categories:\n"
        "%s\n"
        "Input: %s\n\n"
        "Respond in JSON format:\n"
        "{\"category\": \"category_name\", \"confidence\": 0.0-1.0, \"reasoning\": \"explanation\"}",
        categories, input);

    char* response = call_anthropic_api(router->api_key, router->model, prompt, 256);
    if (!response) {
        return NULL;
    }

    ClassificationResult* result = (ClassificationResult*)calloc(1, sizeof(ClassificationResult));
    if (!parse_classification(response, result)) {
        free(result);
        free(response);
        return NULL;
    }

    free(response);
    return result;
}

/**
 * Route an input to the appropriate handler
 */
char* router_route(Router* router, const char* input) {
    ClassificationResult* classification = router_classify(router, input);
    if (!classification) {
        if (router->fallback_handler) {
            return router->fallback_handler(input, router->fallback_user_data);
        }
        return NULL;
    }

    printf("Classification: %s (confidence: %.2f)\n",
           classification->category, classification->confidence);

    // Find matching route
    Route* matching_route = NULL;
    for (size_t i = 0; i < router->route_count; i++) {
        if (strcmp(router->routes[i].category, classification->category) == 0) {
            matching_route = &router->routes[i];
            break;
        }
    }

    char* result = NULL;

    if (matching_route && classification->confidence >= router->confidence_threshold) {
        result = matching_route->handler(input, matching_route->user_data);
    } else if (router->fallback_handler) {
        result = router->fallback_handler(input, router->fallback_user_data);
    }

    free(classification);
    return result;
}

/**
 * Free router resources
 */
void router_free(Router* router) {
    free(router->api_key);
    free(router->model);
    free(router);
}

// Complexity levels for model routing
typedef enum {
    COMPLEXITY_SIMPLE,
    COMPLEXITY_MODERATE,
    COMPLEXITY_COMPLEX
} Complexity;

/**
 * Model router for complexity-based routing
 */
typedef struct ModelRouter {
    char* api_key;
    char* fast_model;
    char* standard_model;
    char* powerful_model;
    char* classification_model;
} ModelRouter;

/**
 * Create model router
 */
ModelRouter* model_router_create(const char* api_key) {
    ModelRouter* router = (ModelRouter*)calloc(1, sizeof(ModelRouter));
    router->api_key = strdup(api_key);
    router->fast_model = strdup("claude-3-haiku-20240307");
    router->standard_model = strdup("claude-sonnet-4-20250514");
    router->powerful_model = strdup("claude-opus-4-20250514");
    router->classification_model = strdup("claude-sonnet-4-20250514");
    return router;
}

/**
 * Assess complexity of input
 */
Complexity model_router_assess(ModelRouter* router, const char* input) {
    char prompt[MAX_INPUT_SIZE];
    snprintf(prompt, sizeof(prompt),
        "Assess the complexity of handling this request:\n\n"
        "%s\n\n"
        "Consider:\n"
        "- simple: Direct factual answers, simple calculations, basic questions\n"
        "- moderate: Analysis, explanations, moderate coding tasks\n"
        "- complex: Deep analysis, complex reasoning, creative writing, complex code\n\n"
        "Respond with just: simple, moderate, or complex",
        input);

    char* response = call_anthropic_api(router->api_key, router->classification_model, prompt, 32);
    if (!response) {
        return COMPLEXITY_MODERATE;
    }

    Complexity result = COMPLEXITY_MODERATE;
    if (strstr(response, "simple")) {
        result = COMPLEXITY_SIMPLE;
    } else if (strstr(response, "complex")) {
        result = COMPLEXITY_COMPLEX;
    }

    free(response);
    return result;
}

/**
 * Route to appropriate model and get response
 */
char* model_router_route(ModelRouter* router, const char* input) {
    Complexity complexity = model_router_assess(router, input);

    const char* model;
    switch (complexity) {
        case COMPLEXITY_SIMPLE:
            model = router->fast_model;
            printf("Using fast model: %s\n", model);
            break;
        case COMPLEXITY_COMPLEX:
            model = router->powerful_model;
            printf("Using powerful model: %s\n", model);
            break;
        default:
            model = router->standard_model;
            printf("Using standard model: %s\n", model);
            break;
    }

    return call_anthropic_api(router->api_key, model, input, 4096);
}

/**
 * Free model router
 */
void model_router_free(ModelRouter* router) {
    free(router->api_key);
    free(router->fast_model);
    free(router->standard_model);
    free(router->powerful_model);
    free(router->classification_model);
    free(router);
}

// Example handlers
char* handle_code_question(const char* input, void* user_data) {
    const char* api_key = (const char*)user_data;
    char prompt[MAX_INPUT_SIZE];
    snprintf(prompt, sizeof(prompt), "As a coding expert, answer: %s", input);
    return call_anthropic_api(api_key, "claude-sonnet-4-20250514", prompt, 4096);
}

char* handle_math_question(const char* input, void* user_data) {
    const char* api_key = (const char*)user_data;
    char prompt[MAX_INPUT_SIZE];
    snprintf(prompt, sizeof(prompt), "As a math expert, solve: %s", input);
    return call_anthropic_api(api_key, "claude-sonnet-4-20250514", prompt, 4096);
}

char* handle_general_question(const char* input, void* user_data) {
    const char* api_key = (const char*)user_data;
    char prompt[MAX_INPUT_SIZE];
    snprintf(prompt, sizeof(prompt), "Answer this question: %s", input);
    return call_anthropic_api(api_key, "claude-sonnet-4-20250514", prompt, 4096);
}

char* handle_fallback(const char* input, void* user_data) {
    printf("Using fallback handler for: %s\n", input);
    return strdup("Handled by fallback");
}

int main() {
    const char* api_key = getenv("ANTHROPIC_API_KEY");
    if (!api_key) {
        fprintf(stderr, "ANTHROPIC_API_KEY environment variable not set\n");
        return 1;
    }

    // Create router
    printf("=== Category Router ===\n");
    Router* router = router_create(api_key, NULL);
    router_set_threshold(router, 0.7);
    router_set_fallback(router, handle_fallback, (void*)api_key);

    // Add routes
    router_add_route(router, "code",
                     "Programming and coding questions",
                     handle_code_question, (void*)api_key);
    router_add_route(router, "math",
                     "Mathematics and calculations",
                     handle_math_question, (void*)api_key);
    router_add_route(router, "general",
                     "General knowledge questions",
                     handle_general_question, (void*)api_key);

    // Route a query
    char* result = router_route(router, "How do I implement a binary search tree?");
    if (result) {
        printf("Result: %s\n", result);
        free(result);
    }

    router_free(router);

    // Model-based routing
    printf("\n=== Model Router ===\n");
    ModelRouter* model_router = model_router_create(api_key);

    char* model_result = model_router_route(model_router, "What is 2+2?");
    if (model_result) {
        printf("Simple query result: %s\n", model_result);
        free(model_result);
    }

    model_result = model_router_route(model_router, "Analyze the complexity of quicksort");
    if (model_result) {
        printf("Complex query result: %s\n", model_result);
        free(model_result);
    }

    model_router_free(model_router);

    return 0;
}

File: skills/agent-patterns/templates/dart/routing.dart
/// Routing Pattern Implementation for Dart
/// Classification-based routing of inputs to specialized handlers

import 'dart:convert';
import 'package:http/http.dart' as http;

/// Anthropic API client
class AnthropicClient {
  final String apiKey;
  final http.Client httpClient;

  AnthropicClient({
    required this.apiKey,
    http.Client? httpClient,
  }) : httpClient = httpClient ?? http.Client();

  Future<String> createMessage({
    required String prompt,
    String model = 'claude-sonnet-4-20250514',
    int maxTokens = 4096,
  }) async {
    final response = await httpClient.post(
      Uri.parse('https://api.anthropic.com/v1/messages'),
      headers: {
        'x-api-key': apiKey,
        'anthropic-version': '2023-06-01',
        'content-type': 'application/json',
      },
      body: jsonEncode({
        'model': model,
        'max_tokens': maxTokens,
        'messages': [
          {'role': 'user', 'content': prompt}
        ],
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('API error (${response.statusCode}): ${response.body}');
    }

    final data = jsonDecode(response.body);
    final content = data['content'] as List;

    for (var block in content) {
      if (block['type'] == 'text') {
        return block['text'] as String;
      }
    }

    throw Exception('No text content in response');
  }
}

/// Result of classification
class ClassificationResult<T> {
  final T category;
  final double confidence;
  final String reasoning;

  ClassificationResult({
    required this.category,
    required this.confidence,
    required this.reasoning,
  });

  @override
  String toString() =>
      'ClassificationResult(category: $category, confidence: $confidence)';
}

/// A route definition
class Route<T, R> {
  final T category;
  final String description;
  final Future<R> Function(String input) handler;

  Route({
    required this.category,
    required this.description,
    required this.handler,
  });
}

/// Routes inputs to specialized handlers based on LLM classification.
///
/// Example:
/// ```dart
/// final router = Router<String, String>(client: client)
///   ..addRoute(Route(
///     category: 'code',
///     description: 'Programming questions',
///     handler: (input) => handleCodeQuestion(input),
///   ))
///   ..addRoute(Route(
///     category: 'general',
///     description: 'General questions',
///     handler: (input) => handleGeneralQuestion(input),
///   ));
///
/// final result = await router.route('How do I sort a list in Python?');
/// ```
class Router<T, R> {
  final AnthropicClient client;
  final String model;
  final double confidenceThreshold;
  final List<Route<T, R>> _routes = [];
  final Future<R> Function(String input)? fallbackHandler;

  Router({
    required this.client,
    this.model = 'claude-sonnet-4-20250514',
    this.confidenceThreshold = 0.7,
    this.fallbackHandler,
  });

  /// Add a route
  Router<T, R> addRoute(Route<T, R> route) {
    _routes.add(route);
    return this;
  }

  /// Classify input into a category
  Future<ClassificationResult<T>> classify(String input) async {
    final categories = _routes.map((r) => '${r.category}: ${r.description}').join('\n');

    final prompt = '''
Classify the following input into one of these categories:
$categories

Input: $input

Respond in JSON format:
{
  "category": "category_name",
  "confidence": 0.0-1.0,
  "reasoning": "explanation"
}
''';

    final response = await client.createMessage(
      prompt: prompt,
      model: model,
      maxTokens: 256,
    );

    final jsonStr = _extractJson(response);
    final data = jsonDecode(jsonStr);

    return ClassificationResult<T>(
      category: data['category'] as T,
      confidence: (data['confidence'] as num).toDouble(),
      reasoning: data['reasoning'] as String,
    );
  }

  /// Route input to appropriate handler
  Future<R> route(String input) async {
    final classification = await classify(input);

    // Find matching route
    final matchingRoute = _routes.cast<Route<T, R>?>().firstWhere(
          (r) => r!.category == classification.category,
          orElse: () => null,
        );

    // Check confidence threshold
    if (matchingRoute != null && classification.confidence >= confidenceThreshold) {
      return matchingRoute.handler(input);
    }

    // Use fallback if available
    if (fallbackHandler != null) {
      return fallbackHandler!(input);
    }

    throw Exception(
      'No handler found for category: ${classification.category} '
      '(confidence: ${classification.confidence})',
    );
  }

  String _extractJson(String text) {
    final start = text.indexOf('{');
    final end = text.lastIndexOf('}');
    if (start >= 0 && end > start) {
      return text.substring(start, end + 1);
    }
    return text;
  }
}

/// Complexity level for model routing
enum Complexity { simple, moderate, complex }

/// Routes inputs to different models based on complexity.
///
/// Example:
/// ```dart
/// final router = ModelRouter(client: client);
/// final result = await router.route('What is 2+2?');  // Uses fast model
/// final result2 = await router.route('Analyze this code...');  // Uses powerful model
/// ```
class ModelRouter {
  final AnthropicClient client;
  final String fastModel;
  final String standardModel;
  final String powerfulModel;
  final String classificationModel;

  ModelRouter({
    required this.client,
    this.fastModel = 'claude-3-haiku-20240307',
    this.standardModel = 'claude-sonnet-4-20250514',
    this.powerfulModel = 'claude-opus-4-20250514',
    this.classificationModel = 'claude-sonnet-4-20250514',
  });

  /// Assess complexity of input
  Future<ClassificationResult<Complexity>> assessComplexity(String input) async {
    final prompt = '''
Assess the complexity of handling this request:

$input

Consider:
- simple: Direct factual answers, simple calculations, basic questions
- moderate: Analysis, explanations, moderate coding tasks
- complex: Deep analysis, complex reasoning, creative writing, complex code

Respond in JSON format:
{
  "complexity": "simple|moderate|complex",
  "confidence": 0.0-1.0,
  "reasoning": "explanation"
}
''';

    final response = await client.createMessage(
      prompt: prompt,
      model: classificationModel,
      maxTokens: 256,
    );

    final jsonStr = _extractJson(response);
    final data = jsonDecode(jsonStr);

    final complexityStr = data['complexity'] as String;
    final complexity = Complexity.values.firstWhere(
      (c) => c.name == complexityStr,
      orElse: () => Complexity.moderate,
    );

    return ClassificationResult<Complexity>(
      category: complexity,
      confidence: (data['confidence'] as num).toDouble(),
      reasoning: data['reasoning'] as String,
    );
  }

  /// Route to appropriate model and get response
  Future<String> route(String input) async {
    final assessment = await assessComplexity(input);

    final model = switch (assessment.category) {
      Complexity.simple => fastModel,
      Complexity.moderate => standardModel,
      Complexity.complex => powerfulModel,
    };

    return client.createMessage(
      prompt: input,
      model: model,
    );
  }

  String _extractJson(String text) {
    final start = text.indexOf('{');
    final end = text.lastIndexOf('}');
    if (start >= 0 && end > start) {
      return text.substring(start, end + 1);
    }
    return text;
  }
}

/// Support ticket for multi-stage routing example
class SupportTicket {
  final String id;
  final String content;
  final String? priority;
  final String? department;
  final String? resolution;

  SupportTicket({
    required this.id,
    required this.content,
    this.priority,
    this.department,
    this.resolution,
  });

  SupportTicket copyWith({
    String? id,
    String? content,
    String? priority,
    String? department,
    String? resolution,
  }) {
    return SupportTicket(
      id: id ?? this.id,
      content: content ?? this.content,
      priority: priority ?? this.priority,
      department: department ?? this.department,
      resolution: resolution ?? this.resolution,
    );
  }
}

/// Multi-stage routing pipeline
class TicketRouter {
  final AnthropicClient client;
  final String model;

  TicketRouter({
    required this.client,
    this.model = 'claude-sonnet-4-20250514',
  });

  Future<SupportTicket> routeTicket(SupportTicket ticket) async {
    // Stage 1: Classify priority
    final priority = await _classifyPriority(ticket.content);

    // Stage 2: Route to department
    final department = await _routeToDepartment(ticket.content);

    // Stage 3: Generate initial response
    final resolution = await _generateResponse(ticket.content, department);

    return ticket.copyWith(
      priority: priority,
      department: department,
      resolution: resolution,
    );
  }

  Future<String> _classifyPriority(String content) async {
    final prompt = '''
Classify the priority of this support ticket:
$content

Options: low, medium, high, critical
Respond with just the priority level.
''';

    final response = await client.createMessage(
      prompt: prompt,
      model: model,
      maxTokens: 32,
    );

    return response.trim().toLowerCase();
  }

  Future<String> _routeToDepartment(String content) async {
    final prompt = '''
Route this support ticket to the appropriate department:
$content

Options: billing, technical, sales, general
Respond with just the department name.
''';

    final response = await client.createMessage(
      prompt: prompt,
      model: model,
      maxTokens: 32,
    );

    return response.trim().toLowerCase();
  }

  Future<String> _generateResponse(String content, String department) async {
    final prompt = '''
As a $department support specialist, provide an initial response to:
$content

Be helpful and professional.
''';

    return client.createMessage(
      prompt: prompt,
      model: model,
    );
  }
}

// Example usage
Future<void> exampleRouting() async {
  final apiKey = const String.fromEnvironment('ANTHROPIC_API_KEY');
  if (apiKey.isEmpty) {
    throw Exception('ANTHROPIC_API_KEY environment variable not set');
  }

  final client = AnthropicClient(apiKey: apiKey);

  // Category-based routing
  final router = Router<String, String>(
    client: client,
    fallbackHandler: (input) async => 'Handling as general query: $input',
  )
    ..addRoute(Route(
      category: 'code',
      description: 'Programming and coding questions',
      handler: (input) async {
        return client.createMessage(
          prompt: 'As a coding expert, answer: $input',
        );
      },
    ))
    ..addRoute(Route(
      category: 'math',
      description: 'Mathematics and calculations',
      handler: (input) async {
        return client.createMessage(
          prompt: 'As a math expert, solve: $input',
        );
      },
    ))
    ..addRoute(Route(
      category: 'general',
      description: 'General knowledge questions',
      handler: (input) async {
        return client.createMessage(
          prompt: 'Answer this question: $input',
        );
      },
    ));

  final result = await router.route('How do I implement a binary search tree?');
  print('Routed result: $result');

  // Model-based routing
  final modelRouter = ModelRouter(client: client);
  final modelResult = await modelRouter.route('What is the capital of France?');
  print('Model routing result: $modelResult');
}

void main() async {
  try {
    await exampleRouting();
  } catch (e) {
    print('Error: $e');
  }
}
